File: cpp/lisp-like/lisp-like/Parser/Parser.cpp
//
//  Parser.cpp
//  lisp-like
//
//  Created by Charles McVicker on 27/01/2018.
//  Copyright © 2018 Charles McVicker. All rights reserved.
//

#include "Parser.hpp"
#include <ctype.h>
#include <iostream>
#include "SymbolExpression.hpp"
#include "ValueExpression.hpp"
#include "FuncCall.hpp"
#include "Dec64.hpp"

void Parser::read(){
    while(!finished()){
        whitespace();
        char cur_char = current();
        if (cur_char == '(') {
            queue.push_back(read_func_call());
        } else if (isdigit(cur_char)){
            queue.push_back(read_num());
        } else if (cur_char == '`'){
            queue.push_back(read_quote());
        } else {
            queue.push_back(read_symbol_expression());
            
        }
    }
}

void Parser::whitespace(){
    while (isspace(current()) or finished()){
        next();
    }
}

char Parser::next(){
    if (text.length() < char_num){
        return text[-1];
    }
    char_num++;
    return text[char_num - 1];
}
char Parser::current(){
    if (text.length() < char_num) {
        return text[-1];
    }
    return text[char_num];
}

bool Parser::finished(){
    return text.length() <= char_num;
}

Expression* Parser::read_func_call(){
    if (current() != '('){
        std::cout << "Syntax error, expected character '\n";
    }
    next();
    whitespace();
    Expression* func;
    if (current() == '('){
        //This is not right, check types in python3
        func = (Func*) read_func_call();
    } else if (isdigit(current())){
        std::cout << "Syntax error, function name cannot start with" << current() << std::endl;
    } else {
        func = read_symbol_expression();
    }
    std::vector<Expression*> args;
    while (next() != ')') {
        whitespace();
        if(current() == '('){
            args.push_back(read_func_call());
        } else if (current() == '`') {
            args.push_back(read_quote());
        } else if(isdigit(current())){
            args.push_back(read_num());
        } else {
            args.push_back(read_symbol_expression());
        }
    }
    return new FuncCall(func, args);
}

Expression* Parser::read_quote(){
    if (current() != '`'){
        next();
    }
    next();
    whitespace();
    if (current() == '(') {
        std::vector<Expression*> contents;
        while (next() != ')'){
            whitespace();
            if(current() == '`') {
                std::cout << "Cannot quote inside of quote\n";
            }
            else if (current() == '('){
                contents.push_back(read_quote());
            } else if (isdigit(current())){
                contents.push_back(read_num());
            } else {
                contents.push_back(read_symbol_expression());
            }
        }
        return new Quote(contents);
    } else {
        std::vector<Expression*> contents;
        if (isdigit(current())) {
            contents.push_back(read_num());
        } else {
            contents.push_back((Expression*)read_symbol());
        }
        return new Quote(contents);
    }
}

Symbol* Parser::read_symbol(){
    if (current() == '('){
        std::cout << "Syntax error, unexpected character ( expected identifier" << current() << "\n";
    }
    std::string identifier = "";
    while (!(isspace(current()) || finished()) && current() != ')') {
        identifier += next();
    }
    return new Symbol(identifier);
}

Expression* Parser::read_symbol_expression(){
    if (current() == '('){
        std::cout << "Syntax error, unexpected character ( expected identifier" << current() << "\n";
    }
    std::string identifier = "";
    while (!(isspace(current()) || finished()) && current() != ')') {
        identifier += next();
    }
    return new SymbolExpression(Symbol(identifier));
}

Expression* Parser::read_num(){
    if (current() == '('){
        std::cout << "Syntax error, unexpected character '('";
    }
    std::string literal = "";
    while (!(isspace(current()) or finished()) && current() != ')'){
        literal += next();
    }
    return new ValueExpression(Value(Dec64(literal)));
}

void Parser::eval(){
    for (int i = 0; i < queue.size(); i++){
        Expression* expr = queue[i];
        if (dynamic_cast<FuncCall*>(expr) or dynamic_cast<SymbolDerefExpression*>(expr)){
            std::cout << "out: " << expr->eval(env).repr() << std::endl;
        } else if(dynamic_cast<Quote*>(expr) or
                  dynamic_cast<SymbolExpression*>(expr)){
            std::cout << "out: " << expr->repr() << std::endl;
        } else {
            std::cout << "out: " << expr->repr() << std::endl;
        }
        delete expr;
    }
}

File: flang/lisp-like/with-quote/parser.py
from state import State
from primatives import Func, FuncCall, Quote


class Parser:
    def __init__(self, text, state=State()):
        self.text = text
        self.state = state
        self.stack = []
        self.errors = []
        self.char_num = 0

    def read(self):
        while not self.finished():
            self.whitespace()
            cur_char = self.current()
            if cur_char == '(':
                self.stack.append(self.read_func_call())
            elif cur_char.isdigit():
                self.stack.append(self.read_int())
            elif cur_char == '`':
                self.stack.append(self.read_quote())
            else:
                self.stack.append(self.read_identifier())

    def read_func_call(self):
        # print('read expr')
        if not self.current() == '(':
            print('Syntax error, expected character \'(\'')
        self.next()
        self.whitespace()
        if self.current() == '(':
            func = self.read_func_call()
        elif self.current().isdigit():
            func = self.read_int()
        else:
            func = self.state.ptr(self.read_identifier())
        args = []
        while self.next() != ')':
            self.whitespace()
            if self.current() == '(':
                args.append(self.read_func_call())
            elif self.current() == '`':
                args.append(self.read_quote())
            # elif self.current() == '"':
            #     args.append(self.read_string())
            elif self.current().isdigit():
                args.append(self.read_int())
            else:
                args.append(self.read_identifier())
        return FuncCall(func, args)

    # def read_string(self):
    #     string = ''
    #     self.next()
    #     while not(self.finished() or self.current() == '"'):
    #         # print('finished?', self.finished())
    #         string += self.current()
    #         if self.current() == '\\':
    #             string += self.next()
    #             self.next()
    #     return string

    def read_quote(self):
        if not self.current() == '`':
            print('Syntax error, expected character \'`\' (back quote)')
        self.next()
        self.whitespace()
        if self.current() == "(":
            contents = []
            while not self.next() == ')':
                self.whitespace()
                if self.current() == '`':
                    contents.append(self.read_quote())
                if self.current() == "(":
                    contents.append(self.read_sub_quote())
                elif self.current().isdigit():
                    contents.append(self.read_int())
                else:
                    contents.append(self.read_identifier())
        else:
            contents = []
            if self.current().isdigit():
                contents.append(self.read_int())
            else:
                contents.append(self.read_identifier())
        return Quote([item for item in contents if not item == ''])

    def read_sub_quote(self):
        if self.current() == "(":
            contents = []
            while not self.next() == ')':
                self.whitespace()
                if self.current() == '`':
                    print("Cannot quote inside of quote, at char {}"
                          .format(self.cur_char))
                if self.current() == "(":
                    contents.append(self.read_sub_quote())
                elif self.current().isdigit():
                    contents.append(self.read_int())
                else:
                    contents.append(self.read_identifier())
        else:
            contents = []
            if self.current().isdigit():
                contents.append(self.read_int())
            else:
                contents.append(self.read_identifier())
        return Quote([item for item in contents if not item == ''])

    def read_var(self):
        # print('read var')
        if self.current() == '(':
            print('Syntax error, unexpected character \'(\',\
                   \nexpected identifier')
        identifier = ''
        while not(self.current().isspace() or self.finished()) and \
                not self.current() == ')':
            # print('finished?', self.finished())
            identifier += self.next()
        return self.state.ptr(identifier)

    def read_identifier(self):
        # print('read var')
        if self.current() == '(':
            print('Syntax error, unexpected character \'(\',\
                   \nexpected identifier')
        identifier = ''
        while not(self.current().isspace() or self.finished()) and \
                not self.current() == ')':
            # print('finished?', self.finished())
            identifier += self.next()
        return identifier

    def read_int(self):
        # print('read int')
        if self.current() == '(':
            print('Syntax error, unexpected character \'(\',\
                   \nexpected identifier')
        literal = ''
        while not(self.current().isspace() or self.finished()) and \
                not self.current() == ')':
            # print('finished?', self.finished())
            literal += self.next()

        return self.state.constant(int(literal))

    def whitespace(self):
        while self.current().isspace() or self.finished():
            # print('nexted in whitespace')
            self.next()

    def next(self):
        if len(self.text) <= self.char_num:
            return self.text[-1]
        char = self.text[self.char_num]
        self.char_num += 1
        return char

    def current(self):
        if len(self.text) <= self.char_num:
            return self.text[-1]
        return self.text[self.char_num]

    def finished(self):
        return len(self.text) <= self.char_num

    def eval(self):
        for expr in self.stack:
            if isinstance(expr, FuncCall):
                # print('got expr')
                print('< <', expr.eval(self.state))
            elif isinstance(expr, Quote):
                print(expr.contents)
            else:
                # print('got var')
                print('< <', self.state.get(expr))

    def error(self):
        return self.errors.pop() if self.errors else None
